Show exception to user in executable version

aldebaran30701 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix`, `hotfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
Add final modifiers to exceptions
Add below line of code where exceptions are commonly thrown as a result of something broken.
`NotifyDescriptor.Exception exceptionDialog = new NotifyDescriptor.Exception(ex);`
this line of code allows the error to be shown to the user in the release version meaning that they will know something is failing rather than failing silently. 

Additions to changelog and whatsnew also added. 

Exception when running from NetBeans:
![image](https://user-images.githubusercontent.com/58677759/124707654-498be680-df38-11eb-8313-66757ab5ad4a.png)

Exception when running from executable:
![build exception](https://user-images.githubusercontent.com/58677759/124707743-6aecd280-df38-11eb-9cc0-f7ee0f95195b.PNG)

Notification bubble is still available to be clicked and therefore possibly reporting the stacktrace to GitHub:
![build exception click bubble](https://user-images.githubusercontent.com/58677759/124707921-a8e9f680-df38-11eb-8afd-d8007ef06883.PNG)

The link within the bubble, when clicked will load the very top window of the first screenshot - this allows the user to see and possibly copy the exception and report it on GitHub.

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
Notification bubble was tested, but a dialog was seemingly the best option as it stops the user continuing without closing the dialog.
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Allows users to see bugs and report them to us on GitHub
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
Allows users to see bugs and report them to us on GitHub
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
None realised
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
Built the executable version in order to test.
Tested using netbeans debug launch.

Both situations I replicated a known bug that threw an exception (or in the case of testing, a bug I planted to replicate the exception).

I ensured that during the executable build, that the exception was shown.
I ensured that during the NetBeans launch, that the exceptions were not extraneous - this doesn't seem to be an issue because one dialog is mentioning the issue and the other one is giving more details. 
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#50 
<!-- Link any applicable issues here -->

aldebaran30701 commented:
Wondering people's thoughts on the sonar issues such as removing the catch block when nothing is done within it. 
Also the unused variable which launches the dialog. If I don't assign it to a variable then NB complains also

antares1470 commented:
> 
> Wondering people's thoughts on the sonar issues such as removing the catch block when nothing is done within it.
> Also the unused variable which launches the dialog. If I don't assign it to a variable then NB complains also

@aldebaran30701 Regarding the catch blocks, my preference is generally to avoid having nothing in a catch block. Looking at the specific instances of this, I think logging the exception should be enough as that alone doesn't halt execution (you need to throw the exception to do that) so the code would otherwise run as it did.

I've encountered the unused variable situation before so understand the dilemma. My thoughts have probably changed over time but currently I would say that the better option of the two is not declare a variable if you don't intend to use it later. We can then mark the subsequent sonar issue as won't do

antares1470 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 
+## Changes in July 2021
+
+-   Exceptions thrown in the `DefaultPluginEnvironment` are now presented to 
```
Are you able to re-words this entry to start with the word "Updated"? Just to align it with other changelog entries

antares1470 commented on CoreGraphNode/src/au/gov/asd/tac/constellation/graph/node/plugins/DefaultPluginEnvironment.java:
```diff
@@ -97,7 +100,8 @@
                 ConstellationLogger.getDefault().pluginStarted(plugin, parameters, graph);
-            } catch (Exception ex) {
+            } catch (final Exception ex) {
+                LOGGER.log(Level.WARNING, EXCEPTION_MESSAGE, ex);
```
Should've mentioned it in my earlier comment but passing the exception as an argument will actually halt the execution (from previous experience) so just omit that so that execution runs as it did. I would also change the warning level to Severe since technically getting here means something went wrong:
`LOGGER.log(Level.SEVERE, EXCEPTION_MESSAGE)`

Similarly for the other log entries you added

aldebaran30701 replied:
So just log it with a default message like "An unexpected exception was caught."
 ?

antares1470 replied:
For now I would maybe log `ex.getLocalisedMessage()`

aldebaran30701 replied:
Done, let me know if that's how you meant

antares1470 replied:
Yep that's what I meant. The only thing I would add is change the log level to Severe rather than Warning. While we're not throwing the exception, getting to those catches does technically mean something went wrong so best to log it as much

arcturus2 replied:
Updated

Auriga2 commented:
I tested this and looks good to me, but the only issue is `org.openide NotifyDescriptor.Exception ` is depricated. Can we use any  alternative suggested below?
http://bits.netbeans.org/dev/javadoc/org-openide-dialogs/org/openide/NotifyDescriptor.html
![image](https://user-images.githubusercontent.com/64826928/125025859-94d20080-e0c2-11eb-9814-7d8bfe6188a0.png)

aldebaran30701 commented:
Good point @Auriga2 
It looks like this post highlights some alternatives.
https://dev.platform.netbeans.narkive.com/VRm1x81G/why-notifydescriptor-exception-is-deprecated#post4

I was working on an alternate design using ErrorManager before trying the current PR. I can switch it over to this if all are in agreeance?

antares1470 commented:
> 
> Good point @Auriga2
> It looks like this post highlights some alternatives.
> https://dev.platform.netbeans.narkive.com/VRm1x81G/why-notifydescriptor-exception-is-deprecated#post4
> 
> I was working on an alternate design using ErrorManager before trying the current PR. I can switch it over to this if all are in agreeance?

If `NotifyDescriptor.Exception` is deprecated then I'm keen to explore other options so happy for you to go with whatever option you think is best @aldebaran30701

arcturus2 commented:
a few things @aldebaran30701:
* agreed also to not use the deprecated method. I had a look at current Apache NetBeans master branch and they still have it deprecated (so in theory we could use it as its not removed) but reading about some of this it looks like there was some deadlock issues
* in case you didn't know a quick way to build the stand alone version is to run the `build-zip` target, then unzip the `constellation.zip` file in the `dist` folder and run it including the path to your jre/jdk like `constellation/bin/constellation --jdkhome <path here>`
* i used the Experimental -> Developer -> Test Parameter Building for testing and introduced an error by changing the value in "Some Integer" from 5 to 5.5 and clicked OK. When run the exe version I would get the error icon appear but no popup. When I tested it through NetBeans I would get a popup every second time. This is so strange but may be clues to a deeper issue. Keen to know if you or @antares1470 or anyone else testing this PR experienced this?

aldebaran30701 commented:
>a few things @aldebaran30701:
agreed also to not use the deprecated method. I had a look at current Apache NetBeans master branch and they still have it deprecated (so in theory we could use it as its not removed) but reading about some of this it looks like there was some deadlock issues

Agreed, plus there appears to be a direct replacement for it that should work fine.

>there is the ConstellationDialog class that you could extend with a ConstellationErrorDialogwith error specific things and use that? also keen to know what you were thinking of using.

I was thinking of just leaving it as is, but I can understand that if we extend the class we only have to change it in one spot in the future. I can add that as a fix to this.

>in case you didn't know a quick way to build the stand alone version is to run the build-zip target, then unzip the constellation.zip file in the dist folder and run it including the path to your jre/jdk like constellation/bin/constellation --jdkhome <path here>
I didn't know that, but will surely be handy for testing futher!


>i used the Experimental -> Developer -> Test Parameter Building for testing and introduced an error by changing the value in "Some Integer" from 5 to 5.5 and clicked OK. When run the exe version I would get the error icon appear but no popup. When I tested it through NetBeans I would get a popup every second time. This is so strange but may be clues to a deeper issue. Keen to know if you or @antares1470 or anyone else testing this PR experienced this?

The problem here is my assumption of things. Because everything in Constellation is a plugin, I added the error handling code to the only place I think I could have caught the application - wide exceptions; the plugin framework. 
It will only report the exception correctly when one occurs within a plugin.

arcturus2 commented:
> It will only report the exception correctly when one occurs within a plugin.
Noted and your approach is fine. The "Test Parameter Building" is a plugin so I would expect it to be caught.

aldebaran30701 commented:
>Noted and your approach is fine. The "Test Parameter Building" is a plugin so I would expect it to be caught.

Just checked the stack trace and I'm wondering if it isn't catching it because it's thrown and handled within JavaFX code. I'm getting only JavaFX libraries within the call stack (NumberFormatException).

aldebaran30701 commented:
Note with this most recent commit, that a heap more has changed. 
I noted that there was some inconsistent behaviour between different plugin types and even the same exceptions across different classes. 

The convention I've went with in my mind is as follows:
- User implements code that may throw unexpected exceptions. These will get caught just inside of the Constellation code. (within methods like run or read of SimpleEditPlugin.java)
- The code will either report to the SimpleEditPlugin.java that it has had an exception occur and therefore cancel, and rethrow the exception for the more generic pluginframework code to handle.
- DefaultPluginEnvironment.java will then catch those exceptions and report them using the PluginInteraction class to display to the user (and also log to the console.)

The thing I was not sure about is the interrupted plugins which occur at the lower levels like in SimpleEditPlugin.java. My implementation takes note that it was interrupted, cancels the plugin via the already used boolean, and reports it up the chain to DefaultPluginEnvironment.java. That class then calls the thread.currentthread.interrupt. 

It seems that during the execution of the application on the build versions, that the LOGGER and also the std.err stream do not get picked up as displaying a dialog box (where it does so in the netbeans launch). I feel this is the case because of using the LOGGER.log(level,message,exception) to print out the stacktrace. The RuntimeExceptionAction simply throws an exception from the toolbar. No caller within Constellation-derived code. This makes it hard to capture in the same way the plugin framework currently uses. I have not found a solution for this.

I've added a class in the testing module which simply adds a toolbar option which throws an exception from within a plugin.

tl;dr - I made the exceptions more consistent within the pluginframework, but still have not got a solution for exceptions thrown outside of it.

antares1470 commented on CoreGraphNode/src/au/gov/asd/tac/constellation/graph/node/plugins/DefaultPluginInteraction.java:
```diff
@@ -222,6 +223,30 @@ public void notify(final PluginNotificationLevel level, final String message) {
     }
 
+    @Override
+    public void notifyException(PluginNotificationLevel level, Exception exception) {
```
final

arcturus2 replied:
Refactored and updated.

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/text/TextPluginInteraction.java:
```diff
@@ -72,4 +72,9 @@ public boolean prompt(String promptName, PluginParameters parameters) {
     }
 
+    @Override
+    public void notifyException(PluginNotificationLevel level, Exception exception) {
```
final

antares1470 commented on CoreTesting/src/au/gov/asd/tac/constellation/testing/RuntimeExceptionPlugin.java:
```diff
@@ -0,0 +1,43 @@
+public class RuntimeExceptionPlugin extends SimpleReadPlugin {
+
+    @Override
+    protected void read(GraphReadMethods graph, PluginInteraction interaction, PluginParameters parameters) throws InterruptedException, PluginException, RuntimeException {
```
final

arcturus2 replied:
Updated

antares1470 commented on CoreTesting/src/au/gov/asd/tac/constellation/testing/RuntimeExceptionPluginAction.java:
```diff
@@ -0,0 +1,54 @@
+
+    private final GraphNode context;
+
+    public RuntimeExceptionPluginAction(GraphNode context) {
```
final

arcturus2 replied:
Updated

arcturus2 commented:
@aldebaran30701 so gave this a test. Just to summarise I am seeing a popup from "Experimental -> Developer -> Throw a RuntimeException from a plugin" but for everything else it is the little icon and no popup shows. Is that consistent to what you are seeing or not? I am testing on Linux.

aldebaran30701 commented:
Converting to draft until I get a few moments to review the comments and make some changes.

arcturus2 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/templates/SimpleEditPlugin.java:
```diff
@@ -116,34 +107,17 @@ public final void run(final PluginGraphs graphs, final PluginInteraction interac
+                    } catch (final Exception ex) {
                         cancelled = true;
-                        LOGGER.log(Level.WARNING, ex, () -> msg0 + SeparatorConstants.SEMICOLON + " " + msg);
-                        throw new RuntimeException(ex);
```
Deciding not to wrap the extension in a RuntimeException() - https://softwareengineering.stackexchange.com/questions/121328/is-it-good-practice-to-catch-a-checked-exception-and-throw-a-runtimeexception

arcturus2 commented:
@aldebaran30701 I closed your PR and started a new one consolidating everything to kick off a fresh review https://github.com/constellation-app/constellation/pull/1459. I hope that is OK with you. I wanted to add some unit tests and then I will open the PR for review. Thanks for starting this work and also for allowing me to continue it.

I really like what its turned into. Check it out when you can.
